File: Code/signal_processing.py
import numpy as np
import scipy.fftpack as fft
import cv2
from scipy.signal import butter, filtfilt
from sklearn.decomposition import FastICA, PCA
from Code.config import LOW_HEART, HIGH_HEART, LOW_RESP, HIGH_RESP, LEVELS, FPS, ALPHA

def build_laplacian_pyramid(frame, levels):
    """Construye una pirámide Laplaciana."""
    pyramid = []
    current = frame.copy()
    for _ in range(levels):
        down = cv2.pyrDown(current)  # Nivel k+1 (Gaussiana)
        up = cv2.pyrUp(down, dstsize=(current.shape[1], current.shape[0]))  # Expandir
        laplacian = cv2.subtract(current, up)  # L_k = G_k - expand(G_{k+1})
        pyramid.append(laplacian)
        current = down  # Continuar con el siguiente nivel Gaussiano
    pyramid.append(current)  # Añadir el último nivel Gaussiano (sin resta)
    return pyramid

def apply_ica_pca(signals):
    """Aplica ICA y PCA a las señales."""
    ica = FastICA(n_components=1, random_state=42)
    pca = PCA(n_components=1)
    return (ica.fit_transform(signals)[:, 0], 
            pca.fit_transform(signals)[:, 0])

def butter_bandpass(lowcut, highcut, fs, order=3):
    """Diseña un filtro de paso de banda Butterworth."""
    nyq = 0.5 * fs
    low = lowcut / nyq
    high = highcut / nyq
    b, a = butter(order, [low, high], btype='band')
    return b, a

def apply_bandpass_filter(signal, lowcut, highcut, fs):
    """Aplica un filtro de paso de banda."""
    b, a = butter_bandpass(lowcut, highcut, fs, order=3)
    return filtfilt(b, a, signal, axis=0)

def estimate_frequency(signal, fps):
    """Estima la frecuencia dominante en una señal."""
    fft_result = np.abs(fft.rfft(signal, axis=0))
    freqs = fft.rfftfreq(len(signal), d=1.0 / fps)
    return freqs[np.argmax(fft_result)] * 60

def process_buffer_evm(frame_buffer):
    """Procesa el buffer de frames usando EVM y extrae signos vitales."""
    pyramids = [build_laplacian_pyramid(frame, LEVELS) for frame in frame_buffer]
    level_signals = np.array([[np.mean(p[level]) for p in pyramids] 
                               for level in range(LEVELS)]).T
    
    heart_signal, resp_signal = apply_ica_pca(level_signals)

    heart_signal = apply_bandpass_filter(heart_signal, LOW_HEART, HIGH_HEART, FPS)
    resp_signal = apply_bandpass_filter(resp_signal, LOW_RESP, HIGH_RESP, FPS) 

    heart_signal *= ALPHA
    resp_signal *= ALPHA

    heart_rate = estimate_frequency(heart_signal, FPS)
    resp_rate = estimate_frequency(resp_signal, FPS)
    
    return heart_rate, resp_rate

File: Cpp/Code/Signal_Processing.cpp
#include "Signal_Processing.h"

// Construcción de pirámide Laplaciana
vector<Mat> buildLaplacianPyramid(const Mat& frame, int levels) {
    vector<Mat> pyramid;
    Mat current = frame.clone();
    for (int i = 0; i < levels; ++i) {
        Mat down, up, lap;
        pyrDown(current, down);
        pyrUp(down, up, current.size());
        subtract(current, up, lap);
        pyramid.push_back(lap);
        current = down;
    }
    pyramid.push_back(current);
    return pyramid;
}

// PCA simple usando Eigen
VectorXd applyPCA(const MatrixXd& data) {
    MatrixXd centered = data.rowwise() - data.colwise().mean();
    MatrixXd cov = (centered.adjoint() * centered) / double(data.rows() - 1);
    SelfAdjointEigenSolver<MatrixXd> eig(cov);
    VectorXd principal = centered * eig.eigenvectors().rightCols(1);
    return principal;
}

// Estimar frecuencia dominante usando FFTW3
double estimateFrequency(const VectorXd& signal, double fps) {
    int N = signal.size();
    double* in = fftw_alloc_real(N);
    fftw_complex* out = fftw_alloc_complex(N / 2 + 1);

    for (int i = 0; i < N; ++i) {
        in[i] = signal(i);
    }

    fftw_plan plan = fftw_plan_dft_r2c_1d(N, in, out, FFTW_ESTIMATE);
    fftw_execute(plan);

    double maxAmp = 0.0;
    int maxIdx = 1; // Ignorar componente DC

    for (int i = 1; i < N / 2 + 1; ++i) {
        double magnitude = sqrt(out[i][0]*out[i][0] + out[i][1]*out[i][1]);
        if (magnitude > maxAmp) {
            maxAmp = magnitude;
            maxIdx = i;
        }
    }

    double freqHz = static_cast<double>(maxIdx) * fps / N;

    fftw_destroy_plan(plan);
    fftw_free(in);
    fftw_free(out);

    return freqHz * 60.0; // BPM
}

// Filtro bandpass tipo Butterworth (simplificado)
VectorXd butterworthBandpassFilter(const VectorXd& input, double lowcut, double highcut, double fs, int order) {
    double nyquist = 0.5 * fs;
    double low = lowcut / nyquist;
    double high = highcut / nyquist;

    VectorXd output = input;
    int N = input.size();
    VectorXd temp(N);

    // Paso alto: media móvil de primer orden
    temp[0] = input[0];
    for (int i = 1; i < N; ++i) {
        temp[i] = input[i] - input[i - 1] + low * temp[i - 1];
    }

    // Paso bajo: suavizado exponencial
    output[0] = temp[0];
    for (int i = 1; i < N; ++i) {
        output[i] = high * temp[i] + (1.0 - high) * output[i - 1];
    }

    return output;
}

// Pipeline principal
pair<double, double> processBufferEVM(const vector<Mat>& frameBuffer) {
    int numFrames = frameBuffer.size();

    // 1. Construir pirámides Laplacianas
    vector<vector<Mat>> pyramids;
    for (const auto& frame : frameBuffer) {
        pyramids.push_back(buildLaplacianPyramid(frame, Config::LEVELS));
    }

    // 2. Extraer señales temporales
    MatrixXd levelSignals(numFrames, Config::LEVELS);
    for (int l = 0; l < Config::LEVELS; ++l) {
        for (int t = 0; t < numFrames; ++t) {
            Scalar meanVal = mean(pyramids[t][l]);
            levelSignals(t, l) = meanVal[0];
        }
    }

    // 3. PCA y amplificación
    VectorXd signal = applyPCA(levelSignals);
    signal *= Config::ALPHA;

    // 4. Filtros de banda para corazón y respiración
    VectorXd heart_filtered = butterworthBandpassFilter(signal, Config::LOW_HEART, Config::HIGH_HEART, Config::FPS);
    VectorXd resp_filtered = butterworthBandpassFilter(signal, Config::LOW_RESP, Config::HIGH_RESP, Config::FPS);

    // 5. Estimación de frecuencia
    double heart_rate = estimateFrequency(heart_filtered, Config::FPS);
    double resp_rate = estimateFrequency(resp_filtered, Config::FPS);

    return {heart_rate, resp_rate};
}
